crates/router/src/connector/paypal.rs

```diff
@@ -460,6 +461,135 @@ impl ConnectorIntegration<api::Authorize, types::PaymentsAuthorizeData, types::P
+                )?)
+                .attach_default_headers()
+                .headers(types::PaymentsPreProcessingType::get_headers(
+                    self, req, connectors,
+                )?)
+                .build(),
+        ))
+    }
+
+    fn handle_response(
+        &self,
+        data: &types::PaymentsPreProcessingRouterData,
+        res: Response,
+    ) -> CustomResult<types::PaymentsPreProcessingRouterData, errors::ConnectorError> {
+        let response: paypal::PaypalPreProcessingResponse = res
+            .response
+            .parse_struct("paypal PaypalPreProcessingResponse")
+            .change_context(errors::ConnectorError::ResponseDeserializationFailed)?;
+
+        // permutation for status to continue payment
+        match (
+            response
+                .payment_source
+                .card
+                .authentication_result
+                .three_d_secure
+                .enrollment_status,
+            response
+                .payment_source
+                .card
+                .authentication_result
+                .three_d_secure
+                .authentication_status,
+            response
+                .payment_source
+                .card
+                .authentication_result
+                .liability_shift,
+        ) {
+            (
+                paypal::EnrollementStatus::Ready,
+                Some(paypal::AuthenticationStatus::Success),
+                paypal::LiabiilityShift::Possible,
+            )
+            | (
+                paypal::EnrollementStatus::Ready,
+                Some(paypal::AuthenticationStatus::Attempted),
+                paypal::LiabiilityShift::Possible,
+            )
+            | (paypal::EnrollementStatus::NotReady, None, paypal::LiabiilityShift::No)
+            | (paypal::EnrollementStatus::Unavailable, None, paypal::LiabiilityShift::No)
+            | (paypal::EnrollementStatus::Bypassed, None, paypal::LiabiilityShift::No) => {
+                Ok(types::PaymentsPreProcessingRouterData {
+                    status: storage_enums::AttemptStatus::AuthenticationSuccessful,
+                    response: Ok(types::PaymentsResponseData::TransactionResponse {
+                        resource_id: types::ResponseId::NoResponseId,
+                        redirection_data: None,
+                        mandate_reference: None,
+                        connector_metadata: None,
+                        network_txn_id: None,
+                        connector_response_reference_id: None,
+                    }),
+                    ..data.clone()
+                })
+            }
+            _ => Ok(types::PaymentsPreProcessingRouterData {
+                response: Err(ErrorResponse {
+                    code: consts::NO_ERROR_CODE.to_string(),
+                    message: consts::NO_ERROR_MESSAGE.to_string(),
+                    reason: Some(consts::CONNECTOR_UNAUTHORIZED_ERROR.to_string()),
```

Comment:
Instead of CONNECTOR_UNAUTHORIZED_ERROR, Create new error which tells the actual reason to fail the payment. "Can not continue with authorization, due to failed liability shift" and also include enrollment_status, authentication_status and liability_shift values in the reason